src/openms/source/FORMAT/OMSFile.cpp

```diff
@@ -0,0 +1,2044 @@
+    // can't both be NULL
+    // @TODO: add constraint that "processing_step_order" must match "..._id"?
+    // @TODO: normalize table? (splitting into multiple tables is awkward here)
+  }
+
+
+  void OMSFile::OMSFileStore::storeAppliedProcessingStep_(
+    const ID::AppliedProcessingStep& step, Size step_order,
+    const String& parent_table, Key parent_id)
+  {
+    // this assumes the "..._AppliedProcessingStep" table exists already!
+    String table = parent_table + "_AppliedProcessingStep";
+
+    QSqlQuery query(QSqlDatabase::database(db_name_));
+    query.prepare("INSERT INTO " + table.toQString() + " VALUES ("  \
+                  ":parent_id, "                                    \
+                  ":processing_step_id, "                           \
+                  ":processing_step_order, "                        \
+                  ":score_type_id, "                                \
+                  ":score)");
+    query.bindValue(":parent_id", parent_id);
+    query.bindValue(":processing_step_order", int(step_order));
+    if (step.processing_step_opt)
+    {
+      query.bindValue(":processing_step_id",
+                      Key(&(**step.processing_step_opt)));
+      if (step.scores.empty()) // insert processing step information only
+      {
+        if (!query.exec())
+        {
+          raiseDBError_(query.lastError(), __LINE__, OPENMS_PRETTY_FUNCTION,
+                        "error inserting data");
+        }
+      }
+    } // else: use NULL for missing processing step reference
+    for (auto score_pair : step.scores)
+    {
+      query.bindValue(":score_type_id", Key(&(*score_pair.first)));
+      query.bindValue(":score", score_pair.second);
+      if (!query.exec())
+      {
+        raiseDBError_(query.lastError(), __LINE__, OPENMS_PRETTY_FUNCTION,
+                      "error inserting data");
+      }
+    }
+  }
+
+
+  void OMSFile::OMSFileStore::storeScoreTypes()
+  {
+    if (id_data_.getScoreTypes().empty()) return;
+
+    createTableCVTerm_();
+    createTable_(
+      "ID_ScoreType",
+      "id INTEGER PRIMARY KEY NOT NULL, "                               \
+      "cv_term_id INTEGER NOT NULL, "                                   \
+      "higher_better NUMERIC NOT NULL CHECK (higher_better in (0, 1)), " \
+      "FOREIGN KEY (cv_term_id) REFERENCES CVTerm (id)");
+
+    QSqlQuery query(QSqlDatabase::database(db_name_));
+    query.prepare("INSERT INTO ID_ScoreType VALUES ("       \
+                  ":id, "                                   \
+                  ":cv_term_id, "                           \
+                  ":higher_better)");
+    for (const ID::ScoreType& score_type : id_data_.getScoreTypes())
+    {
+      Key cv_id = storeCVTerm_(score_type.cv_term);
+      query.bindValue(":id", Key(&score_type)); // use address as primary key
+      query.bindValue(":cv_term_id", cv_id);
+      query.bindValue(":higher_better", int(score_type.higher_better));
+      if (!query.exec())
```

Comment:
@hendrikweisser you dont want to execute a statement here with every loop iteration, you should be able to  execute once at the end of the loop.